cloudinit/signal_handler.py

```diff
@@ -5,28 +5,64 @@
 # Author: Joshua Harlow <harlowja@yahoo-inc.com>
 #
 # This file is part of cloud-init. See LICENSE file for license information.
+import contextlib
 import inspect
+import threading
 import logging
 import signal
 import sys
+import types
 from io import StringIO
+from typing import Callable, Dict, Final, NamedTuple, Union
 
 from cloudinit import version as vr
 from cloudinit.log import log_util
 
 LOG = logging.getLogger(__name__)
 
-
-BACK_FRAME_TRACE_DEPTH = 3
-EXIT_FOR = {
-    signal.SIGINT: ("Cloud-init %(version)s received SIGINT, exiting...", 1),
-    signal.SIGTERM: ("Cloud-init %(version)s received SIGTERM, exiting...", 1),
-    # Can't be caught...
-    # signal.SIGKILL: ('Cloud-init killed, exiting...', 1),
-    signal.SIGABRT: ("Cloud-init %(version)s received SIGABRT, exiting...", 1),
+SIG_MESSAGE: Final = "Cloud-init {} received {}, exiting\n"
+BACK_FRAME_TRACE_DEPTH: Final = 3
+SIGNALS: Final[Dict[int, str]] = {
+    signal.SIGINT: "Cloud-init %(version)s received SIGINT, exiting",
+    signal.SIGTERM: "Cloud-init %(version)s received SIGTERM, exiting",
+    signal.SIGABRT: "Cloud-init %(version)s received SIGABRT, exiting",
 }
 
 
+class ExitBehavior(NamedTuple):
+    exit_code: int
+    log_level: int
+
+
+SIGNAL_EXIT_BEHAVIOR_CRASH: Final = ExitBehavior(1, logging.ERROR)
+SIGNAL_EXIT_BEHAVIOR_QUIET: Final = ExitBehavior(0, logging.INFO)
+_SIGNAL_EXIT_BEHAVIOR = SIGNAL_EXIT_BEHAVIOR_CRASH
+_SUSPEND_WRITE_LOCK = threading.RLock()
+
+
+def inspect_handler(sig: Union[int, Callable, None]) -> None:
+    """inspect_handler() logs signal handler state"""
+    if callable(sig):
+        # only produce a log when the signal handler isn't in the expected
+        # default state
+        if not isinstance(sig, types.BuiltinFunctionType):
+            LOG.info("Signal state [%s] - previously custom handler.", sig)
+    elif sig == signal.SIG_IGN:
+        LOG.info("Signal state [SIG_IGN] - previously ignored.")
+    elif sig is None:
+        LOG.info("Signal state [None] - previously not installed from Python.")
+    elif sig != signal.SIG_DFL:
```

Comment:
```suggestion
    elif sig == signal.SIG_DFL:
```

I think this is the root of all of the current integration test failures on thsi branch.